### ISSUE TITLE
How is the HTTPS forwarded?

### ISSUE THREAD

**Issue by mtekman** [tsani/jerrington.me#12]
In regards to your "Self hosted ngrok alternative" post: https://jerrington.me/posts/2019-01-29-self-hosted-ngrok.html

```
server {
    server_name tunnel.yourdomain;

    access_log /var/log/nginx/$host;
    
    # These three lines are new.
    listen 443 ssl;
    ssl_certificate /path/to/tls/cert/fullchain.pem;
    ssl_certificate_key /path/to/tls/cert/privkey.pem;

    location / {
	    proxy_pass http://localhost:3333/;
	    proxy_set_header X-Real-IP $remote_addr;
	    proxy_set_header Host $host;
	    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto https;
	    proxy_redirect off;
    }

    error_page 502 /50x.html;
    location = /50x.html {
	    root /usr/share/nginx/html;
    }
}
```

This is perhaps a bit of a naive question, but if someone connects to my publically accessible device over HTTPS (443), how is the content forwarded to my private webserver?

I understand how requests on the default http port(s) are forward to port 3333, which is picked up by the HTTP port on private server (8888). 

But how does this work for HTTPS?

Is port 443 also redirected into port 3333, and then transferred to port  8888 (HTTP?) on my private device? 

Do I not need to (ssh -R) bind 443 on the proxy to my private device's HTTPS port (that isn't 8888)?

**Comment by tsani** [comment]
Hi @mtekman, thank you for reaching out and asking. When you use `listen 443 ssl` in your nginx configuration block for the server, you're instructing nginx to listen on port 443 and to do TLS termination. If someone connects to your public server `tunnel.yourdomain` (running this nginx config) on port 443 they will "be in that server block" so to speak. The request they make will be forwarded to `http://localhost:3333` as written in the `location` block.

In short, there's no big difference between doing the forwarding with and without encryption. The only difference is that nginx will deal with terminate TLS for you and forward the unencrypted HTTP through your (encrypted) SSH tunnel listening on port 3333.

If you were to reverse tunnel port 443 itself, then you would bypass nginx entirely, and the raw HTTPS traffic would be forwarded through the SSH tunnel to your private server.

**Comment by mtekman** [comment]
@tsani thanks for your reply! I think I see, and have a working solution based on your excellent post.

One thing I'm wondering about is, where should I install the Lets Encrypt certificate? The way I see it, based on your reply above is:
* If 443 terminates on the public proxy host, then the certificate should be installed there.
* If 443 is forwarded back to the private server (via port 3333), then the certificate should be installed on the private server.

Is that right?

**Comment by tsani** [comment]
@mtekman yes exactly, that's right.

The best practice would be to terminate TLS on the public host, and install the certificate there. TLS certificates are tied to domain names, and your public host would be the one reachable via your domain name. I would find it kind of weird to install a certificate for the public domain on the private host that isn't reachable via that domain except via the tunnel.

**Comment by mtekman** [comment]
Makes sense, thanks!
